### PR TITLE
Add a WithOS annotation.

### PR DESCRIPTION
**Pull request by abayer**
This is basically a copy of WithOS from the ATH (along with the
RuleAnnotation class it needs) - I've hit cases where I'd like unit
tests to be able to run on some OSes but be skipped on others without
the tests actually failing. The immediate example of this is trying to
build plugins that have unit tests using Docker Pipeline while on OS
X, because ow, lots of problems (like
https://issues.jenkins-ci.org/browse/JENKINS-36802). Fun!

cc @reviewbybees

### CODE REVIEW

File: src/main/java/org/jvnet/hudson/test/junit/RuleAnnotation.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * <p>
+ * This allows annotations on test class/method to add additional setup/shutdown behaviours.
+ *
+ * @author Kohsuke Kawaguchi
+ */
+@Retention(RUNTIME)
+@Target(ANNOTATION_TYPE)
+@Documented
+public @interface RuleAnnotation {
```
Comment (jglick) [inline src/main/java/org/jvnet/hudson/test/junit/RuleAnnotation.java]: I do not see anything actually looking for this. Am I missing something?

**Comment by jglick** [comment]
I tend to find it simpler and clearer to explicitly throw `AssumptionViolatedException` from the test. That way it is obvious how to tweak the condition, to look for a particular binary for example.

File: src/main/java/org/jvnet/hudson/test/junit/WithOS.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    public class RuleImpl implements TestRule {
+        @Override
+        public Statement apply(final Statement base, final Description d) {
+            return new Statement() {
+
+                @Override
+                public void evaluate() throws Throwable {
+                    Class<?> testSuite = d.getTestClass();
+                    check(d.getAnnotation(WithOS.class), testSuite);
+                    check(testSuite.getAnnotation(WithOS.class), testSuite);
```
Comment (stephenc) [inline src/main/java/org/jvnet/hudson/test/junit/WithOS.java]: Why is this necessary, the annotation should be present in the `Description` otherwise there is a bug in JUnit IIUC

**Comment by stephenc** [comment]
If we drop the TestSuite hack I think it is OK as that way the annotation is in the expected places

File: src/main/java/org/jvnet/hudson/test/junit/RuleAnnotation.java (hunk 1 above)
Comment (abayer) [inline src/main/java/org/jvnet/hudson/test/junit/RuleAnnotation.java]: `@RuleAnnotation` is on `WithOS`.

File: src/main/java/org/jvnet/hudson/test/junit/WithOS.java (hunk 2 above)
Comment (abayer) [inline src/main/java/org/jvnet/hudson/test/junit/WithOS.java]: Honestly, that's copy-pasted from the ATH, so it may not be needed at all. I'll experiment with it.

**Comment by abayer** [comment]
@jglick - so the reason I want this rather than throwing an Exception is that I don't want the tests to fail when their assumptions aren't satisfied - whatever gets me that result (skipped rather than error/failed) is fine by me.

**Comment by stephenc** [comment]
@abayer `assumeThat(os.isWindows()` would get you that...

but a more interesting PoV for me on this annotation is that we could index it to find the list of tests that are OS specific and then mark those as needing to have at least one run on a matrix of OSs...

Now that is not something we can do right now, but having the annotation in place to let us start capturing that information is a big +1 in my view

File: src/main/java/org/jvnet/hudson/test/junit/RuleAnnotation.java (hunk 1 above)
Comment (jglick) [inline src/main/java/org/jvnet/hudson/test/junit/RuleAnnotation.java]: I see where the annotation is _applied_ of course. But where is it _consulted_?

Comment (jglick) [inline src/main/java/org/jvnet/hudson/test/junit/RuleAnnotation.java]: [reference](https://github.com/jenkinsci/acceptance-test-harness/blob/c35e9b940864f2ce76debc45c9712e36c1bbb644/src/main/java/org/jenkinsci/test/acceptance/junit/JenkinsAcceptanceTestRule.java#L99)

File: src/main/java/org/jvnet/hudson/test/junit/RuleAnnotation.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    /**
+     * The rule class that defines the setup/shutdown behaviour.
+     *
+     * The instance is obtained through Guice.
```
Comment (jglick) [inline src/main/java/org/jvnet/hudson/test/junit/RuleAnnotation.java]: Huh??

File: src/main/java/org/jvnet/hudson/test/junit/WithOS.java (hunk 2 above)
Comment (jglick) [inline src/main/java/org/jvnet/hudson/test/junit/WithOS.java]: This is to support annotations on both the test case (method) or suite (class).

**Comment by jglick** [comment]
> I don't want the tests to fail when their assumptions aren't satisfied - whatever gets me that result (skipped rather than error/failed) is fine by me

That is what throwing `AssumptionViolatedException` from a test accomplishes. Now if @stephenc wants to construct a system like he proposes, great, but for now that remains hypothetical.

At any rate, :bug: since as written it does not do anything.

File: src/main/java/org/jvnet/hudson/test/junit/RuleAnnotation.java (hunk 3 above)
Comment (stephenc) [inline src/main/java/org/jvnet/hudson/test/junit/RuleAnnotation.java]: Oh dear me more Guice "magic"... I say rip that out, the original author uses it like goto
